server_udp.go

```diff
@@ -42,83 +46,169 @@ func (w *UDPResponseWriter) WriteMsg(m *dns.Msg) error {
+
+	// Mark the server as started.
+	// Note that we don't check if it was started before as
+	// Serve* methods can be called multiple times.
+	s.started = true
+
+	// Track an active UDP listener
+	s.udpListeners[l] = struct{}{}
+	return err
+}
+
+// cleanUpUDP waits until all UDP messages before cleaning up
+func (s *Server) cleanUpUDP(udpWg *sync.WaitGroup, l *net.UDPConn) {
+	// Wait until UDP messages are processed
+	udpWg.Wait()
+
+	// Not using it anymore so can be removed from the active listeners
+	s.lock.Lock()
+	delete(s.udpListeners, l)
+	s.lock.Unlock()
+
+	// The work is finished
+	s.wg.Done()
+}
+
+// readUDPMsg reads incoming UDP message
+func (s *Server) readUDPMsg(l *net.UDPConn) ([]byte, *dns.SessionUDP, error) {
+	_ = l.SetReadDeadline(time.Now().Add(defaultReadTimeout))
+	b := make([]byte, dns.MinMsgSize)
+	n, sess, err := dns.ReadFromSessionUDP(l, b)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return b[:n], sess, err
+}
+
+// serveUDPMsg handles incoming DNS message
+func (s *Server) serveUDPMsg(b []byte, certTxt string, sess *dns.SessionUDP, l *net.UDPConn) {
+	if bytes.Equal(b[:clientMagicSize], s.ResolverCert.ClientMagic[:]) {
+		// This is an encrypted message, we should decrypt it
+		m, q, err := s.decrypt(b)
+		if err == nil {
+			rw := &UDPResponseWriter{
+				udpConn: l,
+				sess:    sess,
+				encrypt: s.encrypt,
+				req:     m,
+				query:   q,
 			}
-		} else {
-			// Most likely this a DNS message requesting the certificate
-			reply, err := s.handleHandshake(b, certTxt)
+			err = s.serveDNS(rw, m)
 			if err != nil {
-				log.Tracef("Failed to process a plain DNS query: %v", err)
-			}
-			if err == nil {
-				_, _ = l.WriteTo(reply, addr)
+				log.Tracef("failed to process a DNS query: %v", err)
 			}
+		} else {
+			log.Tracef("failed to decrypt incoming message len=%d: %v", len(b), err)
 		}
-
+	} else {
+		// Most likely this a DNS message requesting the certificate
+		reply, err := s.handleHandshake(b, certTxt)
 		if err != nil {
-			if isConnClosed(err) {
-				log.Info("udpListen.ReadFrom() returned because we're reading from a closed connection, exiting loop")
-			} else {
-				log.Info("got error when reading from UDP listen: %s", err)
-			}
-			break
+			log.Tracef("failed to process a plain DNS query: %v", err)
+		}
+		if err == nil {
+			// Ignore errors, we don't care and can't handle them anyway
+			_, _ = dns.WriteToSessionUDP(l, reply, sess)
 		}
 	}
+}
+
+// setUDPSocketOptions method is necessary to be able to use dns.ReadFromSessionUDP / dns.WriteToSessionUDP
+func setUDPSocketOptions(conn *net.UDPConn) error {
+	if runtime.GOOS == "windows" {
+		return nil
+	}
 
+	// Try setting the flags for both families and ignore the errors unless they
+	// both error.
+	err6 := ipv6.NewPacketConn(conn).SetControlMessage(ipv6.FlagDst|ipv6.FlagInterface, true)
+	err4 := ipv4.NewPacketConn(conn).SetControlMessage(ipv4.FlagDst|ipv4.FlagInterface, true)
+	if err6 != nil && err4 != nil {
+		return err4
```

Comment:
I don't quite get this.  Perhaps it's better to just return the first encountered error, as usual?